### ISSUE TITLE
Breaking changes related to infinite generators missing in TF 2.2 release notes

### ISSUE THREAD

**Issue by bersbersbers** [tensorflow/tensorflow#39199]
## URL(s) with the issue:

Please provide a link to the documentation entry, for example:

https://github.com/tensorflow/tensorflow/releases
https://www.tensorflow.org/api_docs/python/tf/keras/Model?version=nightly

I have spent the better part of a working day on tracking down several infinite loops using TF 2.2.0. In summary, the following two pieces of code run fine in TF 2.1.0, and run in infinite loops in TF 2.2.0:

## Issue 1

```
"""Bug 1."""
import numpy as np
from tensorflow.keras.applications import vgg16
from tensorflow.keras.preprocessing.image import ImageDataGenerator

N = 9
data = ImageDataGenerator().flow(np.empty((N, 224, 224, 3)), np.empty((N, 1000)))

model = vgg16.VGG16()
model.compile(optimizer="adam", loss="categorical_crossentropy")
model.fit(
    x=data,
    epochs=2,
    verbose=2,  # 1
    # steps_per_epoch=N,
)

# https://www.tensorflow.org/api_docs/python/tf/keras/Model?version=nightly
# "When passing an infinitely repeating dataset, you must specify the steps_per_epoch argument."
# This is a breaking change that should be part of the release notes.
```
This code auto-selects `steps_per_epoch=1` ("Train for 1 steps") in TF 2.1.0, which is correct (default `batch_size` is 32, `len(train_flow)` is 1). In TF 2.2.0, it's an infinite loop that is hard to debug. After setting `verbose=1`, one can see that something is in fact happening, with the number of steps per epoch growing without limits; from there, looking at the docs, one notices that `steps_per_epoch` is required ("When passing an infinitely repeating dataset, you must specify the steps_per_epoch argument."). But it's hard to recognize that breaking change from the release notes. (It's even hard to know that the `ImageDataGenerator().flow()` does indeed repeat infinitely, since it does have a length.)

## Issues 2 and 3

```
"""Bug 2."""
import numpy as np
from tensorflow.keras.applications import vgg16
from tensorflow.keras.preprocessing.image import ImageDataGenerator

N = 9
data = ImageDataGenerator().flow(np.empty((N, 224, 224, 3)), np.empty((N, 1000)))

model = vgg16.VGG16()
model.compile(optimizer="adam", loss="categorical_crossentropy")
model.fit(
    x=data,
    epochs=2,
    verbose=1,
    # steps_per_epoch=N,
    validation_data=data,
    ## validation_steps=N,
)

# https://www.tensorflow.org/api_docs/python/tf/keras/Model?version=nightly
# "Note that `validation_data` does not support all the data types that are supported in `x`, eg, dict, generator [...]"
# This is a breaking change that should be part of the release notes.
```
Similarly, this code auto-selects `steps_per_epoch=1` and `validation_steps=1` ("Train for 1 steps, validate for 1 steps") in TF 2.1.0, which is correct (default `batch_size` is 32, `len(val_flow)` is 1). In TF 2.2.0, it's again an infinite loop that is hard to debug, even if one sets `verbose=1` **and** `steps_per_epoch=N` (since there is no verbosity on the validation part). Again, looking at the docs, one notices that `validation_steps` is required ("In the case of an infinitely repeated dataset, it will run into an infinite loop."). But, again, it's hard to recognize that breaking change from the release notes. (Again, it's even hard to know that the `ImageDataGenerator().flow()` does indeed repeat infinitely, since it does have a length.)

Finally, the nightly docs say that "`validation_data` does not support all the data types that are supported in `x`, eg, dict, generator [...]". As you can see above, using a `ImageDataGenerator` still works if you adhere to the other requirements. If use of any generator is disencouraged from now on, this is a breaking change that needs to go into the release notes in my opinion.

**Comment by Saduf2019** [comment]
I am able to replicate this issue even on gpu, please find the gist of [nightly](https://colab.sandbox.google.com/gist/Saduf2019/02c85eb6a8a1dca5aa731be6a4512f81/nightly_gpu.ipynb) and [2.1](https://colab.sandbox.google.com/gist/Saduf2019/d228e57a1c59c9bd04e45dd7c36489d5/gpu.ipynb).

**Comment by jgciudad** [comment]
Same issue here regarding use of generators for validation data. If we're working with generators, how are we supposed to check validation accuracy and loss? It doesn't have any sense to include the `'validation_split'` argument to create two subsets in `ImageDataGenerator`, but then allow to use just the training one in `fit`.

**Comment by jgciudad** [comment]
Same issue here regarding use of generators for validation data. If we're working with generators, how are we supposed to check validation accuracy and loss? It doesn't have any sense to include the `'validation_split'` argument to create two subsets in `ImageDataGenerator`, but then allow to use just the training one in `fit`.

**Comment by jvishnuvardhan** [comment]
@bersbersbers I ran your codes in `TF2.2` and `tf-nightly` and both are not any infinite loop now. Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/d181ea48127195b8d037fdeb112b13a4/nightly_gpu.ipynb). 

Please verify once and close the issue if this was resolved for you. Thanks!

**Comment by goldiegadde** [comment]
This is similiar to #39277 and has been fixed in #b53ed4d. The fix is available in nightly and the next TF version release will have it.
